### DISCUSSION TITLE
Change locale during runtime

### DISCUSSION THREAD

**Discussion by xDivisionByZerox** [faker-js/faker#1049] [Polls]
Localization is a hard topic. But to make data as individual as possible, Faker provides many locales you can choose from for your data generation.
But changing the locale during runtime makes a lot of trouble for the faker-js team. And we couldn't come up with a real use case for this. So we discussed a possible solution and came to the conclusion of dropping local changes during runtime and instead need to provide them during Faker instance initialization.

Our library wouldn't be the same without YOU (our community), so we would like your feedback on our proposal.

**Comment by SimeonC** [comment]
I'd voted yes as I remembered using the locale function for storybook stories. Our app is in 18 languages so it is helpful for mocking user data in the correct language.

Though after I voted I realised that as long as we could still use multiple locales at the same time we wouldn't need the dynamic locale switching.

Something like `import fakerEn from '.../en'; import fakerJa from '.../ja';` would probably be enough. I do remember moment time zone does the same thing, if really necessary this approach does allow code splitting as well.

**Reply by xDivisionByZerox** [reply]
This would be exactly how you would work with multiple locales in the future. Instead of changing the locale, you would instantiate multiple faker instances with their according locale.

**Reply by pkuczynski** [reply]
User who needs two different languages at the same time could instantiate two different faker instances:

```js
import { faker: en } from '@faker/faker-js'
import { faker: de } from '@faker/faker-js/de'

export const faker = { en, de }

faker.en.name.firstName()
faker.de.name.firstName()
```

**Comment by JessicaSachs** [comment]
Yeah so long as a dynamic import can serve the additional locales on demand, ripping it out by default would be A++.

**Comment by pkuczynski** [comment]
If we keep supporting dynamic switching (https://github.com/faker-js/faker/pull/642), things like this would be possible:
        
```js
import { faker } from '@faker/faker-js' // faker logic
import { de } from '@faker/faker-js/locales/de' // only definitions

faker.locale = de
faker.firstName()

faker.locale = null
faker.firstName()    // en
```

* pros
    * user can add his own locale definitions if we publish data types
    * we could expose locales as individual faker packages, which user can install - this would limit the main package size significantly
    * allows user to manually override certain data sets or extend when we don't have them yet 
* cons
  * publishing types for definitions is quite limiting as we do not always have an easy way of marking them as deprecated + we would need to document them (extra work)
  * when you provide an incomplete package how do we fall back? maybe always to `en` which means `faker` comes always with `en`
  * do users even need adding custom definitions? probably not. faker meant to be simple, not complex.

**Reply by xDivisionByZerox** [reply]
What you propose there is not "keeping" but refactoring the behavior. Currently, you instantiate a faker instance and can only switch to locales already defined. So what you do is not bind the locale to a complete definition but to a string that represents that definition `faker.locale = 'en' `.

**Reply by pkuczynski** [reply]
Yes yes, I know. I am trying to bring different options to the table. And still think publishing locales as separate modules is a nice idea, as it would limit the size of a package. And most of the people would not need all the locales at the same time I guess?

**Reply by pkuczynski** [reply]
I created a separate side discussion about publishing separate packages for every locale: https://github.com/faker-js/faker/discussions/1199

**Comment by pkuczynski** [comment]
Another interesting concept is having the ability to generate random data across multiple locales randomly. Idea (stolen from https://github.com/joke2k/faker#localization):

```js
import { faker } from '@faker/faker-js' // faker logic
import { en } from '@faker/faker-js/locales/en' // only definitions
import { it } from '@faker/faker-js/locales/it' // only definitions
import { jp } from '@faker/faker-js/locales/jp' // only definitions

faker.locales = [it, en, jp]

for (const i = 0; i < 10; i++) {
  console.log(faker.name.fullName())
}

// 鈴木 陽一
// Leslie Moreno
// Emma Williams
// 渡辺 裕美子
// Marcantonio Galuppi
// Martha Davis
// Kristen Turner
// 中津川 春香
// Ashley Castillo
// 山田 桃子
```

Would anybody need this?

**Reply by pkuczynski** [reply]
This idea has been moved to a separate pool https://github.com/faker-js/faker/discussions/1197
